### PR TITLE
feat(rpc) implemented v0.5.1 missing fields + bumped starknet.rs to match

### PR DESCRIPTION
**Pull request by antiyro**
### feat(rpc) implemented v0.5.1 missing fields + bumped starknet.rs to match

This pull request is to update the Madara code base to the latest version of RPC specs v0.5.1 as defined in here [starknet api json](https://github.com/starkware-libs/starknet-specs/blob/v0.5.1/api/starknet_api_openrpc.json). To achieve this, it was necessary to bump [starknet.rs](starknet.rs) version to [64ebc36](https://github.com/xJonathanLEI/starknet-rs/commit/64ebc364c0c346e81b715c5b4a3b32ef37b055c8) commit in starknet-api and blockifier.

The fields `l1_gas_price`, `message_hash`, `execution_resources` has been defaulted since they're beeing implemented [here](https://github.com/keep-starknet-strange/madara/issues/1303)

This pull request does not implement any changes but is needed to solve the [v0.5.1](https://github.com/keep-starknet-strange/madara/issues/1291) madara rpc upgrade

### CODE REVIEW

**Comment by antiyro** [comment]
Some problems in this PR are still to be solved because of new fields/impl in starknet.rs:

- `FunctionInvocation`(`InvokeTransactionTrace`): needs a refacto fo `EventContent` and `MessageToL1`to `OrderedEvent` and `OrderedMessage`
- `state_diff`field missing in Transaction traces field (invoke, l1message, etc)

Implies changes in starknet-api.

More to be identified.

**Comment by antiyro** [comment]
https://github.com/keep-starknet-strange/madara/issues/1327
https://github.com/keep-starknet-strange/madara/issues/1326
https://github.com/keep-starknet-strange/madara/issues/1325

Should make this PR mergeable

File: crates/primitives/transactions/src/execution.rs [hunk 1]
```diff
@@ -83,6 +86,47 @@ impl SimulateTxVersionOffset for TransactionVersion {
...
+    /// The number of ecdsa builtin instances
+    pub ecdsa_builtin_applications: u64,
+    /// The number of bitwise builtin instances
+    pub bitwise_builtin_applications: u64,
+    /// The number of keccak builtin instances
+    pub keccak_builtin_applications: u64,
+}
+
+impl StarknetRPCExecutionResources {
+    pub fn into_core_execution_resources(self) -> CoreExecutionResources {
```
Comment (tdelabro) [inline crates/primitives/transactions/src/execution.rs:115]: Why not use `From` trait?

File: crates/pallets/starknet/runtime_api/Cargo.toml [hunk 2]
```diff
@@ -7,6 +7,11 @@ edition = "2021"
 
 [dependencies]
 # Madara primitives
+mp-block = { workspace = true, features = [
+  "parity-scale-codec",
+  "serde",
+  "scale-info",
+] }
```
Comment (tdelabro) [inline crates/pallets/starknet/runtime_api/Cargo.toml:14]: Why is this now required if no changes have been made to `runtime_api/src/lib.rs` ?

File: crates/pallets/starknet/src/blockifier_state_adapter.rs [hunk 3]
```diff
@@ -88,7 +88,7 @@ impl<T: Config> State for BlockifierStateAdapter<T> {
 
     fn increment_nonce(&mut self, contract_address: ContractAddress) -> StateResult<()> {
         let current_nonce = Pallet::<T>::nonce(contract_address);
-        let current_nonce: FieldElement = current_nonce.0.into();
+        let current_nonce: FieldElement = Felt252Wrapper::from(current_nonce.0).into();
```
Comment (tdelabro) [inline crates/pallets/starknet/src/blockifier_state_adapter.rs:91]: Why this change? The previous `From` impl does not work anymore?

File: crates/pallets/starknet/src/lib.rs [hunk 4]
```diff
@@ -1179,14 +1191,15 @@ impl<T: Config> Pallet<T> {
                     results.push(SimulatedTransaction {
                         transaction_trace,
                         fee_estimation: FeeEstimate { gas_consumed, gas_price, overall_fee },
-                    })
+                    });
                 }
                 Err(e) => {
-                    log::error!("Failed to simulate transactions: {:?}, error: {:?}", tx, e);
+                    log::error!("Failed to simulate transaction: {:?}, error: {:?}", tx, e);
```
Comment (tdelabro) [inline crates/pallets/starknet/src/lib.rs:1197]: Why do we log here if we also return the error? Is it just for debugging?

I don't think we want our node to log an error everytime an use simulate an invalid tx

File: crates/primitives/block/Cargo.toml [hunk 5]
```diff
@@ -12,6 +12,11 @@ targets = ["x86_64-unknown-linux-gnu"]
```
Comment (tdelabro) [inline crates/primitives/block/Cargo.toml:12]: You must be the first person I review who does the "features thing" correctly! GG

File: crates/primitives/block/src/header.rs [hunk 6]
```diff
@@ -101,7 +106,6 @@ impl Header {
             self.event_count.into(),
             self.event_commitment.into(),
             self.protocol_version.into(),
-            Felt252Wrapper::ZERO,
```
Comment (tdelabro) [inline crates/primitives/block/src/header.rs:104]: https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/header/#block_hash

Are you sure about that?
Can you link the source you use as a reference for this impl?

File: crates/primitives/block/src/lib.rs [hunk 7]
```diff
@@ -13,10 +13,35 @@ use mp_felt::Felt252Wrapper;
 use mp_hashers::HasherT;
 use mp_transactions::compute_hash::ComputeTransactionHash;
 use mp_transactions::Transaction;
+use serde::{Deserialize, Serialize};
```
Comment (tdelabro) [inline crates/primitives/block/src/lib.rs:16]: Create a `serde` feature, and derive `serde` in the same way `parity-scale-codec` is

File: crates/primitives/block/src/tests.rs [hunk 8]
```diff
@@ -2,79 +2,78 @@ use core::convert::TryFrom;
...
+        block_number.into(),
+        global_state_root.into(),
+        sequencer_address.into(),
+        block_timestamp.into(),
+        transaction_count.into(),
+        transaction_commitment.into(),
+        event_count.into(),
+        event_commitment.into(),
+        protocol_version.into(),
+        Felt252Wrapper::ZERO,
```
Comment (tdelabro) [inline crates/primitives/block/src/tests.rs:68]: Especially when you are keeping the zero here

File: crates/primitives/commitments/Cargo.toml [hunk 9]
```diff
@@ -17,6 +17,7 @@ mp-felt = { workspace = true }
 mp-hashers = { workspace = true }
 mp-transactions = { workspace = true }
 starknet-crypto = { workspace = true }
+starknet-ff = { workspace = true }
```
Comment (tdelabro) [inline crates/primitives/commitments/Cargo.toml:20]: Why is it not optional anymore?
The changes in `crates/primitives/commitments/src/lib.rs` don't use it. Do they?

File: crates/primitives/fee/src/lib.rs [hunk 10]
```diff
@@ -22,13 +22,15 @@ use blockifier::transaction::transaction_utils::{calculate_l1_gas_usage, calcula
 #[cfg(not(feature = "std"))]
 use hashbrown::HashMap;
 use mp_state::StateChanges;
+use serde::Deserialize;
```
Comment (tdelabro) [inline crates/primitives/fee/src/lib.rs:25]: Isn't `serde` optional?
I think if you `cd crates/primitives/fee && cargo check` a compilation error will be raised

File: crates/primitives/fee/src/lib.rs [hunk 11]
```diff
@@ -53,6 +55,24 @@ pub const TRANSFER_SELECTOR_HASH: [u8; 32] = [
     64, 236, 165, 180, 130, 209, 46,
 ]; // starknet_keccak(TRANSFER_SELECTOR_NAME.as_bytes()).to_le_bytes();
 
+#[serde_with::serde_as]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Deserialize)] // Added Deserialize here
```
Comment (tdelabro) [inline crates/primitives/fee/src/lib.rs:59]: What is this comment for?

File: crates/primitives/fee/src/lib.rs [hunk 12]
```diff
@@ -53,6 +55,24 @@ pub const TRANSFER_SELECTOR_HASH: [u8; 32] = [
...
+#[cfg_attr(feature = "parity-scale-codec", derive(parity_scale_codec::Encode, parity_scale_codec::Decode))]
+#[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
```
Comment (tdelabro) [inline crates/primitives/fee/src/lib.rs:62]: Deserailize should be there too

File: crates/primitives/simulations/Cargo.toml [hunk 13]
```diff
@@ -15,8 +15,22 @@ serde = { workspace = true, features = ["derive"], optional = true }
...
+mp-state = { workspace = true, features = [
+  "parity-scale-codec",
+  "serde",
+  "scale-info",
+] }
+mp-transactions = { workspace = true, features = [
+  "parity-scale-codec",
+  "serde",
+  "scale-info",
+] }
```
Comment (tdelabro) [inline crates/primitives/simulations/Cargo.toml:33]: serde is an optional dependency here.
Same for psc and si.

So you don't have to build with those features by default. They are only build if the corresponding feature is enabled on this crate

Comment (tdelabro) [reply crates/primitives/simulations/Cargo.toml:33]: Add mp-state/<serde/psc/si> in the features bellow. Same for mp-transaction

File: crates/primitives/state/Cargo.toml [hunk 14]
```diff
@@ -12,8 +12,34 @@ targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
 blockifier = { workspace = true }
+mp-felt = { workspace = true, features = [
+  "parity-scale-codec",
+  "serde",
+  "scale-info",
```
Comment (tdelabro) [inline crates/primitives/state/Cargo.toml:18]: same, they go bellow, in the features

File: crates/primitives/state/Cargo.toml [hunk 15]
```diff
@@ -12,8 +12,34 @@ targets = ["x86_64-unknown-linux-gnu"]
...
+] }
+serde = { workspace = true, optional = true }
+serde_with = { workspace = true, optional = true }
```
Comment (tdelabro) [inline crates/primitives/state/Cargo.toml:21]: I like to visually separate the optional deps from the other ones with a
```rust
# Optiona
```
comment

It helps readability.


True for all you Cargo.toml

File: crates/primitives/state/src/lib.rs [hunk 16]
```diff
@@ -1,16 +1,141 @@
...
+}
+
+/// The change in state applied in this block, given as a mapping of addresses to the new values
+/// and/or new contracts.
+#[serde_with::serde_as]
+#[derive(Default, Debug, Clone)]
+#[cfg_attr(feature = "parity-scale-codec", derive(parity_scale_codec::Encode, parity_scale_codec::Decode))]
+#[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+pub struct StateDiff {
```
Comment (tdelabro) [inline crates/primitives/state/src/lib.rs:123]: Are all those new types only used for the RPC? Never for internal computation?

If so don't put them here. Put them all together in some RPC-type crate.

I ask because we already use a bunch of different types internally for modeling state diff. So it's weird to add another here but not use it in the crate

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 17]
```diff
@@ -2,6 +2,7 @@
 
 ## Next release
 
+- feat(rpc): Added state diff field in trace fields
```
[the remark this answers is not in the document]
Comment (antiyro) [reply CHANGELOG.md]: Added more detailed changelogs

File: crates/client/rpc-core/starknet_openRPC.json [hunk 18]
```diff
@@ -1,12 +1,26 @@
 {
   "openrpc": "1.0.0-rc1",
   "info": {
-    "version": "0.50.0",
-    "title": "Starknet Node API",
+    "version": "0.5.1",
+    "title": "StarkNet Node API",
```
[the remark this answers is not in the document]
Comment (antiyro) [reply crates/client/rpc-core/starknet_openRPC.json]: Just checked. We can remove it. Removed it.

File: crates/primitives/transactions/src/execution.rs (hunk 1 above)
Comment (antiyro) [reply crates/primitives/transactions/src/execution.rs]: It was handled by an external contributor, I corrected it.

File: crates/pallets/starknet/runtime_api/Cargo.toml (hunk 2 above)
Comment (antiyro) [reply crates/pallets/starknet/runtime_api/Cargo.toml]: Removed

File: crates/pallets/starknet/src/blockifier_state_adapter.rs (hunk 3 above)
Comment (antiyro) [reply crates/pallets/starknet/src/blockifier_state_adapter.rs:91]: Yes for some reason `FieldElement` from `StarkFelt` cannot be performed anymore. My fix is to go trough the wrapper.

File: crates/pallets/starknet/src/lib.rs (hunk 4 above)
Comment (antiyro) [reply crates/pallets/starknet/src/lib.rs]: Removed log here

File: crates/primitives/block/Cargo.toml (hunk 5 above)
Comment (antiyro) [reply crates/primitives/block/Cargo.toml:12]: lfg!

File: crates/primitives/block/src/header.rs (hunk 6 above)
Comment (antiyro) [reply crates/primitives/block/src/header.rs:104]: My apologies this is a mistake I did when correcting the corresponding tests. Added back.

File: crates/primitives/block/src/lib.rs (hunk 7 above)
Comment (antiyro) [reply crates/primitives/block/src/lib.rs:16]: Done

File: crates/primitives/block/src/tests.rs (hunk 8 above)
Comment (antiyro) [reply crates/primitives/block/src/tests.rs:68]: Yes I problably confuse tests with impl

File: crates/primitives/commitments/Cargo.toml (hunk 9 above)
Comment (antiyro) [reply crates/primitives/commitments/Cargo.toml]: Corrected it

File: crates/primitives/fee/src/lib.rs (hunk 10 above)
Comment (antiyro) [reply crates/primitives/fee/src/lib.rs]: Corrected it

File: crates/primitives/fee/src/lib.rs (hunk 11 above)
Comment (antiyro) [reply crates/primitives/fee/src/lib.rs]: Personal note, removed it

File: crates/primitives/fee/src/lib.rs (hunk 12 above)
Comment (antiyro) [reply crates/primitives/fee/src/lib.rs]: Corrected it.

File: crates/primitives/simulations/Cargo.toml (hunk 13 above)
Comment (antiyro) [reply crates/primitives/simulations/Cargo.toml]: Yes I forgot thanks

File: crates/primitives/state/Cargo.toml (hunk 14 above)
Comment (antiyro) [reply crates/primitives/state/Cargo.toml:18]: Correct me if I'm wrong but seems that I only forgot for scale-info here

File: crates/primitives/state/Cargo.toml (hunk 15 above)
Comment (antiyro) [reply crates/primitives/state/Cargo.toml]: Corrected for all the referenced cargos (removing Substrate dependencies comment to avoid having multiple comment lines)

File: crates/primitives/state/src/lib.rs (hunk 16 above)
Comment (antiyro) [reply crates/primitives/state/src/lib.rs:123]: Were would you like me to do so? rpc/types?

File: crates/primitives/block/src/lib.rs (hunk 7 above)
Comment (antiyro) [reply crates/primitives/block/src/lib.rs:16]: Actually I've removed optional feature for serde in Block as it is always needed for rename.

File: crates/primitives/block/src/header.rs (hunk 6 above)
Comment (tdelabro) [reply crates/primitives/block/src/header.rs:104]: you added it back in the last position when it was in the penultimate one before.

File: crates/primitives/block/src/lib.rs (hunk 7 above)
Comment (tdelabro) [reply crates/primitives/block/src/lib.rs:16]: I don't think that it is how it works.

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub enum BlockTag {
    #[cfg_attr(feature = "serde", serde(rename = "latest"))]
    Latest,
    #[cfg_attr(feature = "serde", serde(rename = "pending"))]
    Pending,
}
```

And "serde" should still be optional, and should be behind a feature flag

File: crates/primitives/commitments/Cargo.toml [hunk 19]
```diff
@@ -43,4 +43,4 @@ std = [
 ]
 parity-scale-codec = ["dep:parity-scale-codec", "mp-felt/parity-scale-codec"]
 scale-info = ["dep:scale-info", "mp-felt/scale-info"]
-serde = ["dep:serde", "dep:starknet-ff", "starknet-ff/serde", "mp-felt/serde"]
```
Comment (tdelabro) [inline crates/primitives/commitments/Cargo.toml:46]: I don't think this is good.

serde = ["dep:serde", "dep:starknet-ff", "starknet-ff?/serde", "mp-felt/serde"]

Comment (tdelabro) [reply crates/primitives/commitments/Cargo.toml:46]: But we are going to nuke this anyway. Replace it by our bonzai impl

File: crates/primitives/simulations/Cargo.toml (hunk 13 above)
Comment (tdelabro) [reply crates/primitives/simulations/Cargo.toml]: What I meant is that you should replace this:
```
mp-transactions = { workspace = true, features = [
  "parity-scale-codec",
  "serde",
  "scale-info",
] }
```

by this
```
mp-transactions = { workspace = true }
```
 
 as long as you don
 ```
 parity-scale-codec = [
   ...,
   "mp-transactions/parity-scale-codec",
]
```

File: crates/primitives/state/Cargo.toml (hunk 14 above)
Comment (tdelabro) [reply crates/primitives/state/Cargo.toml:18]: Same as above, you can remove ` features = [
  "parity-scale-codec",
  "serde",
  "scale-info",
  ]` as it is managed below in the `features` section

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Review by tdelabro — APPROVED** [review]
lfg
